Update IronRure Battery package to latest source and move to use Github Actions pipelines

oscarjaergren opened the pull request:
As well as many updates to code style, latest source generator instead of DLLImport for improve code.

oscarjaergren committed:
Update project files and improve code structure

- Enhanced `.editorconfig` with C# style preferences.
- Updated `IronRure.sln` for Visual Studio version and project renaming.
- Changed `Alice.csproj`, `filtering.csproj`, and `grep.csproj` to target .NET 9.0.
- Refactored `AliceBenchTask.cs` and `Program.cs` for better readability.
- Simplified class definitions in `CaptureNamesEnumerable.cs` and `CaptureNamesEnumerator.cs`.
- Improved overall code structure and maintainability across the `IronRure` project.
- Updated test files for better structure and consistency.
- Added `MemoryLeakTests.cs` to check for potential memory leaks.

oscarjaergren committed:
Clean up

oscarjaergren committed:
Update CI/CD workflows and project configurations

- Added GitHub Actions workflows for CI and publishing.
- Updated `GitVersion.yml` to use `main` as the default branch.
- Removed `azure-pipelines.yml` in favor of GitHub Actions.
- Updated project files to reference new package versions.
- Implemented proper `IDisposable` pattern in `Options.cs`.
- Enhanced memory leak tests with increased size parameters.

oscarjaergren committed:
Refactor CI workflows and update solution name

Updated `ci.yml` and `publish.yml` for improved formatting and structure. Changed solution name from `IronRe2.sln` to `IronRure.sln`. Added new NuGet source for `oscarjaergren` and updated build, test, and packaging steps. Commented out unnecessary lines for clarity and future modifications.

oscarjaergren committed:
Comment out GitHub Packages source for crispthinking

The GitHub Packages NuGet source for the `crispthinking` repository has been commented out in `ci.yml`. The original line was removed, and a commented version was added for reference. The line for the `oscarjaergren` repository remains unchanged.

oscarjaergren committed:
Improve resource disposal safety and fix comments

Added null-conditional operators to safely dispose of resources in the CaptureNamesEnumerator, Captures, Options, Regex, RegexIter, and RegexSet classes, preventing potential NullReferenceExceptions. Updated comments for clarity and corrected spelling in the RegexIter class.

oscarjaergren committed:
Remove unused usings

oscarjaergren committed:
Enhance CI/CD workflows for multi-platform support

Updated `ci.yml` to allow workflow calls, expanded job matrix to include Windows and macOS, and modified permissions for package writing.

Restructured `publish.yml` to create a new "Publish IronRure" workflow that triggers on GitHub releases, incorporating a build job from `ci.yml` and a publish job for NuGet packages.

oscarjaergren committed:
Add CI steps for packing and uploading NuGet packages

This commit introduces two new steps in the `ci.yml` file for the `ubuntu-latest` environment. The "Pack" step runs the `dotnet pack` command to create NuGet packages from the solution without rebuilding it, outputting the artifacts to a specified directory. The "Upload NuGet packages" step uses the `actions/upload-artifact` action to upload the generated NuGet packages from the `Artifacts` directory for later access in the workflow.

oscarjaergren committed:
Merge pull request #1 from oscarjaergren/feature/update-regex

Feature/update regex

oscarjaergren committed:
Update version NR

oscarjaergren committed:
Fix path

oscarjaergren committed:
Always include the DLL's and TRY To fix path for sure

oscarjaergren committed:
Fix publish attempt too many to count

abbottdev commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,75 @@
+      #   run: dotnet nuget add source --username ${{ github.actor }} --password ${{ secrets.GITHUB_TOKEN }} --store-password-in-clear-text --name github "https://nuget.pkg.github.com/crispthinking/index.json"
+
+      - name: Add GitHub Packages NuGet Source
+        run: dotnet nuget add source --username ${{ github.actor }} --password ${{ secrets.GITHUB_TOKEN }} --store-password-in-clear-text --name github "https://nuget.pkg.github.com/oscarjaergren/index.json"
```
Don't think this package feed is quite right 😄

iwillspeak replied:
I don't think we want _either_ of these pre-release sources right?

oscarjaergren replied:
Dropped, but can't merge in until IronRure package battery been sorted so I can update the version properly.

abbottdev commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,75 @@
+          dotnet-version: '9.0.x'
+    
+      # - name: Add GitHub Packages NuGet Source
+      #   run: dotnet nuget add source --username ${{ github.actor }} --password ${{ secrets.GITHUB_TOKEN }} --store-password-in-clear-text --name github "https://nuget.pkg.github.com/crispthinking/index.json"
```
drop comments

oscarjaergren replied:
Dropped, but can't merge in until IronRure package battery been sorted so I can update the version properly as they come from different sources now. Also this is the IwillSpeak feed, so probably needs to get sorted to be from CrispThinking without that

iwillspeak commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,31 @@
+          name: artifacts
+          path: bin/artifacts/
+
+      - name: Publish NuGet Packages to GitHub Packages
```
Are we missing a publish to big boy Nuget?

oscarjaergren replied:
Added

iwillspeak commented on src/IronRure/Regex.cs:
```diff
@@ -1,277 +1,293 @@
+    /// <summary>
+    ///   Finalizer
+    /// </summary>
+    ~Regex()
```
The safe handle has the finaliser. I don't think we should make regex finalisable.

oscarjaergren replied:
That one should be sorted now.

iwillspeak commented on src/IronRure/RegexSet.cs:
```diff
@@ -3,131 +3,150 @@
+    /// <summary>
+    ///   Finalizer to ensure resources are released.
+    /// </summary>
+    ~RegexSet()
```
I don't think we directly have any unmanaged resources here? WHy the finaliser?

iwillspeak replied:
I think the old dispose impl was correct.

oscarjaergren replied:
Reverted

oscarjaergren replied:
Or rather, improved rather than reverted.

oscarjaergren commented:
Sure we should merge into IWillSpeak and not CrispThinking?

oscarjaergren committed:
Remove GitHub Packages NuGet source steps in CI

This commit removes the steps in the `ci.yml` file that added GitHub Packages as a NuGet source for the `crispthinking` and `oscarjaergren` repositories. The remaining CI steps include setting up .NET, restoring dependencies, and building the solution.

oscarjaergren committed:
Update NuGet publishing to target NuGet.org

Removed GitHub Packages publishing step and added a new
step to publish NuGet packages to NuGet.org. Updated
the API key source to use the `NUGET_APIKEY` secret
and added comments for clarity on handling duplicates
and workflow retries.

oscarjaergren committed:
Enable nullable and remove Finalizer from RegExSet

oscarjaergren committed:
Run linter

oscarjaergren committed:
Make nullable

oscarjaergren committed:
Fix typos

oscarjaergren committed:
Make CaptureNamesEnum nullable

oscarjaergren committed:
RegEx

oscarjaergren committed:
Don't use primary ctor

oscarjaergren committed:
Skip nullable, it's a PITA

oscarjaergren committed:
Simplify

oscarjaergren committed:
Disable more null

oscarjaergren committed:
Refactor Regex and Captures classes for improved safety

- Changed nullable Raw properties to non-nullable in Captures and Regex classes.
- Updated error handling in Regex to provide default messages.
- Added null checks for Raw property usage in matching methods.
- Simplified Captures method to directly set Matched property.
- Changed CaptureAll return types to ensure non-null Captures.
- Cleaned up disposal logic in RegexIter by removing unnecessary checks.
- Updated RegexSet to use non-nullable handle type.
- Improved clarity and correctness in RegexTests, including fixing method names.

iwillspeak commented on src/IronRure/Captures.cs:
```diff
@@ -2,78 +2,111 @@
+    /// <summary>
+    ///     The raw, unmanaged, handle to the captures group
+    /// </summary>
+    public CapturesHandle Raw { get; }
```
Probably should be internal you know

oscarjaergren replied:
Adjusted

iwillspeak commented on src/IronRure/Captures.cs:
```diff
@@ -2,78 +2,111 @@
+/// <summary>
+///     A set of captures for a given regex.
+/// </summary>
+public class Captures : IDisposable, IEnumerable<Match>
```
We can probably just mark `Captures` as sealed right and avoid the need for a finaliser.

oscarjaergren replied:
Adjusted

iwillspeak commented on src/IronRure/Captures.cs:
```diff
@@ -2,78 +2,111 @@
+    /// </summary>
+    ~Captures()
+    {
+        Dispose(false);
```
again, not sure we need a finaliser here. We're not holding any unmanaged resources directly here so if we're not disposed the undelrying handles will call their own finalisers when they become garbage.

oscarjaergren replied:
Adjusted

oscarjaergren committed:
Ensure we use the latest version

iwillspeak commented on src/IronRure/Options.cs:
```diff
@@ -1,38 +1,80 @@
+    /// </summary>
+    ~Options()
+    {
+        Dispose(false);
```
Redundant finaliser.

oscarjaergren replied:
Adjusted

iwillspeak commented on src/IronRure/IronRure.csproj:
```diff
@@ -1,21 +1,18 @@
+		<AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+	</PropertyGroup>
+	<ItemGroup>
+		<PackageReference Include="IronRure.Batteries-Darwin" Version="154.1.1" />
```
1111.0.5

iwillspeak replied:
scratch that. .7

oscarjaergren replied:
Adjusted

oscarjaergren committed:
Update version

oscarjaergren committed:
Improve options and remove finalizer

oscarjaergren committed:
Make captures sealed, remove finalizer and other clean up

oscarjaergren committed:
Mark internal

oscarjaergren committed:
Fix typo

iwillspeak commented on src/IronRure/CapturesIter.cs:
```diff
@@ -2,51 +2,64 @@
+    /// <inheritdoc />
+    public void Reset()
+    {
+        throw new NotImplementedException();
```
II know this isn’t you. But pretty sure this should be not supported rather than not implemented.

oscarjaergren replied:
Adjusted it.

iwillspeak commented on src/IronRure/Options.cs:
```diff
@@ -1,38 +1,61 @@
+        set
         {
-            set => RureFfi.rure_options_size_limit(Raw, new UIntPtr(value));
+            if (_disposed)
```
I think the throw if is preferred. Keeps the throw out of the current method which reduces code size of the method, means throws are less likely to be branch predicted, and leads to better instruction cache usage.

oscarjaergren replied:
I think I've sorted it to what you meant by changing it to  ObjectDisposedException.ThrowIf(_disposed, nameof(Options));

iwillspeak commented on src/IronRure/Options.cs:
```diff
@@ -1,38 +1,61 @@
+            return;
         }
+
+        Raw?.Dispose();
```
I don’t think the handles can ever be null in these dispose calls can they?

oscarjaergren replied:
In regards to this, Im not sure, if I went all the way down and enabled nullable it would suggest so (which I tried and gave up on cause It required me to think too much about nullability, the source library and how we want to handle null, this is technically just a leftover off that, so I have reverted it.

oscarjaergren committed:
Change to throwif and remove uneeded null check

oscarjaergren committed:
Update Reset method to throw NotSupportedException

Modified the `Reset` method in the `CapturesIter` class to throw a `NotSupportedException` instead of a `NotImplementedException`. This change clarifies that the `Reset` method is not supported in this context.
